### PR TITLE
Scraper

### PR DESCRIPTION
**Pull request by morlick**

### CODE REVIEW

File: bfex/components/scrapper/run.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+from scrapper import *
+
+# this should be done with a real service but
+# I do not know how that will work yet
+data = json.load(open(r'../../../../faculty.json'))
```
Comment (GarryBullock) [inline bfex/components/scrapper/run.py]: This data will be available in elastic search under the faculty index. But you wont need it if we implement the scraper as designed. It shouldn't know about the faculty at all. Only urls.

File: bfex/components/scrapper/run.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+for item in data:
+    name_from_json = item['name']
+    name_from_json = re.sub('[.]', '', name_from_json)
+    name_list = re.findall('[A-Z][^A-Z]*', name_from_json)
+    formated_name = "-".join(name_list).lower()
```
Comment (GarryBullock) [inline bfex/components/scrapper/run.py]: This code (that generates names) should go under something like bfex/common/utils.py. In reality its useful everywhere, and we may reuse it.

File: bfex/components/scrapper/scrapper.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        self.text = text
+
+    def set_date(self, date):
+        self.date = date
+
+    def add_meta(self, hash, meta):
+        self.meta_data[hash] = meta
+
+
+class Scraper(object):
```
Comment (GarryBullock) [inline bfex/components/scrapper/scrapper.py]: Scraper should be an abstract base class (import abc). We shouldn't ever instantiate it, as it doesn't provide any methods. Its essentially an interface for what methods need to be implemented by each scraper.

File: bfex/components/scrapper/run.py (hunk 1 above)
Comment (MandyMeindersma) [reply bfex/components/scrapper/run.py]: yes but currently this is the only way to get any urls

**Comment by GarryBullock** [comment]
Also, please rename the folder structure -> scraper. Its gonna drive me nuts

File: bfex/components/scrapper/run.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+    # concatenate names into url
+    url = base_url + formated_name
+
+    ret = ScraperFactory.create_scraper(ScraperFactory, url, PROFILE)
```
[the remark this answers is not in the document]
Comment (GarryBullock) [reply bfex/components/scrapper/run.py]: If you make ScraperFactory create_scraper a static method, you dont need to pass in self. It doesn't make sense to as is.

File: bfex/components/scrapper/scrapper.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    def set_title(self, title):
+        self.title = title
+
+    def set_text(self, text):
+        self.text = text
+
+    def set_date(self, date):
+        self.date = date
+
+    def add_meta(self, hash, meta):
```
Comment (GarryBullock) [inline bfex/components/scrapper/scrapper.py]: Hash is confusing, as you aren't currently hashing things into it. Youre just passing a string. Maybe call it attr ?

File: bfex/components/scrapper/scrapper.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        soup = self.get_url()
+        links = soup.find_all("a")
+        for link in links:
+            try:
+                if 'orcid' in link.attrs['href']:
+                    #print(formated_name)
+                    #print("    "+"ORCID ID")
+                    #print("    "+link.attrs['href'])
+                    scrapp = Scrapp()
+                    scrapp.add_meta("orcid_link", link.attrs['href'])
```
Comment (GarryBullock) [inline bfex/components/scrapper/scrapper.py]: The orcid and research id should be in the same scrapp, not 2 different ones. A scrapp should deal with one distinct piece of data. In this case, it is the faculty-directory page.

**Comment by GarryBullock** [comment]
I also agree with Mandy that we should use the builtin enum class to handle our link types.

File: bfex/components/scrapper/scrapper.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+PROFILE = 2
+
+class Scrapp(object):
+    def __init__(self):
+        self.title = ""
+        self.text = ""
+        self.date = ""
+        self.meta_data = {}
+
+    def set_title(self, title):
```
Comment (MandyMeindersma) [inline bfex/components/scrapper/scrapper.py]: in regards to our slack convo. A scrapp should have a `self.formatted_name` or something, you don't have to code how we get it now but we have all this info about a prof and not their name attached to it.
we should also have a `self.data_source`

File: bfex/components/scraper/scraper.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+class Scraper(ABC):
+    
+    def __init__(self, url, type):
+        self.url = url
+        self.type = type
+        
+    def validate_url(self):
+        pass
+
+    def get_url(self):
```
Comment (MandyMeindersma) [inline bfex/components/scraper/scraper.py]: Sorry one more comment :(

this isn't getting the url right? its getting the content?

I think we should make @GarryBullock  change his uml cause `get_content` would return the content of the html and `get_scrapp` should return the srapp. He has it different in uml

https://github.com/CMPUT401-Find-an-Expert/Find-an-Expert/wiki/Technical-Documentation

**Comment by MandyMeindersma** [comment]
looks great :100: I just have one more small change mostly for garry

**Comment by MandyMeindersma** [comment]
Imma merge this. Cool?

**Comment by GarryBullock** [comment]
Looks good mostly. Lets stick to naming schemes with underscores delimiting words. ie. bfex/blueprints/data_ingestion.py. Its just cleaner overall. I know im being nitpicky at this point but its nice to have consistent formatting.

**Pull request merged**